test/framework-ondata-mocha.js

```diff
@@ -0,0 +1,526 @@
+
+            // Define the fRunUpdater that framework should invoke
+            var fRunUpdater = function() { 
+                // Verify this is vni context
+                this.should.have.all.keys( 
+                    'delete', 'inputStates', 'errorState', 'outputState', 'node');
+                
+                test.verifyState( this.inputStates( 'input1' ), '', input1 );
+                test.verifyState( this.inputStates( 'input2' ), '', input2 );
+
+                // update state
+                var state = this.outputState();
+                state.data = executedFRunUpdater;
+                state.lm = createLm();
+                this.outputState( state );
+            }
+
+            // Create a noflo component and get the node instance for it
+            var wrapper = { fRunUpdater: fRunUpdater };
+            var node = test.createComponent( 
+                factory({
+                          description: "Test Description",
+                          inPorts: {
+                              input1: {
+                                  datatype: 'string',
+                                  description: "a string port",
+                                  required: true
+                              },
+                              input2: {
+                                  datatype: 'string',
+                                  description: "a string port",
+                                  required: true
+                              }
+                          }
+                        }, 
+                        wrapper )
+            );
+
+            node.vnis.should.be.an( 'object' );
+            Object.keys( node.vnis ).should.have.length(0);
+
+            // Send data to the input port and verify that the fRunUpdater function is called.
+            // We use a promise here because this section is asynchronous
+            var socket1;
+            var socket2;
+            return new Promise( function(done, fail) { 
+
+                test.onOutPortData(node, 'output', done);
+                test.onOutPortData(node, 'error', fail);
+
+                // Attach both ports up front 
+                var socket1 = noflo.internalSocket.createSocket();
+                node._component_under_test.inPorts['input1'].attach(socket1);
+                var socket2 = noflo.internalSocket.createSocket();
+                node._component_under_test.inPorts['input2'].attach(socket2);
+
+                socket1.send(input1);
+                socket2.send(input2);
+
+                socket1.disconnect();
+                socket2.disconnect();
+
+            }).then( function( done ) { 
+               // Verify we got the output state we expect
+               done.vnid.should.equal( '' ); 
+               done.data.should.equal( executedFRunUpdater ); 
+
+               // Cleanup 
+               test.detachAllInputSockets( node );
+            }, function( fail ) { 
+               assert.isNotOk( fail );
```

Comment:
Since the test is returning a promise, there is no need to assert that it didn't fail. The test will fail if the promise is rejected.